RecoPixelVertexing/PixelTriplets/plugins/CAHitNtupletGeneratorKernels.cu

```diff
@@ -1,5 +1,132 @@
+  grid.sync();
+  countOrFillHitInTracks<CountOrFill::count>(tuples_d, quality_d, tuple_d);
+  grid.sync();
+  finalizeCoop(view, ws);
+  grid.sync();
+  countOrFillHitInTracks<CountOrFill::fill>(tuples_d, quality_d, tuple_d);
+}
+
+__inline__ void populateHitInTracks(HitContainer const *tuples_d,
+                                    Quality const *quality_d,
+                                    HitToTuple::View view,
+                                    cudaStream_t cudaStream) {
+  using View = HitToTuple::View;
+  int blockSize = 64;
+  int nblocks = nQuadrupletBlocks(blockSize);
+
+  auto kernel = kernel_populateHitInTracks;
+
+  assert(nblocks > 0);
+  auto nOnes = view.size();
+  auto nchunks = nOnes / blockSize + 1;
+  auto ws = cms::cuda::make_device_unique<View::Counter[]>(nchunks, cudaStream);
+  auto wsp = ws.get();
+  // FIXME: discuss with FW team: cuda calls are expensive and not needed for each event
+  static int const maxBlocks = maxCoopBlocks(kernel, blockSize, 0, 0);
+  auto ncoopblocks = std::min(nblocks, maxBlocks);
+  assert(ncoopblocks > 0);
+  void *kernelArgs[] = {&tuples_d, &quality_d, &view, &wsp};
+  dim3 dimBlock(blockSize, 1, 1);
+  dim3 dimGrid(ncoopblocks, 1, 1);
+  // launch
+  cudaCheck(cudaLaunchCooperativeKernel((void *)kernel, dimGrid, dimBlock, kernelArgs, 0, cudaStream));
+}
+
+__global__ void kernel_populateMultiplicity(HitContainer const *__restrict__ tuples_d,
+                                            Quality const *__restrict__ quality_d,
+                                            caConstants::TupleMultiplicity::View view,
+                                            caConstants::TupleMultiplicity::View::Counter *ws) {
+  namespace cg = cooperative_groups;
+  auto grid = cg::this_grid();
+  auto tupleMultiplicity_d = static_cast<caConstants::TupleMultiplicity *>(view.assoc);
+  zeroAndInitCoop(view);
+  grid.sync();
+  countOrFillMultiplicity<CountOrFill::count>(tuples_d, quality_d, tupleMultiplicity_d);
+  grid.sync();
+  finalizeCoop(view, ws);
+  grid.sync();
+  countOrFillMultiplicity<CountOrFill::fill>(tuples_d, quality_d, tupleMultiplicity_d);
+}
+
+__inline__ void populateMultiplicity(HitContainer const *tuples_d,
+                                     Quality const *quality_d,
+                                     caConstants::TupleMultiplicity *tupleMultiplicity_d,
+                                     cudaStream_t cudaStream) {
+  auto kernel = kernel_populateMultiplicity;
+  using View = caConstants::TupleMultiplicity::View;
+  View view = {tupleMultiplicity_d, nullptr, nullptr, -1, -1};
+
+  int blockSize = 128;
+  int nblocks = (3 * caConstants::maxTuples / 4 + blockSize - 1) / blockSize;
+  assert(nblocks > 0);
+  auto nOnes = view.size();
+  auto nchunks = nOnes / blockSize + 1;
+  auto ws = cms::cuda::make_device_unique<View::Counter[]>(nchunks, cudaStream);
+  auto wsp = ws.get();
+  // FIXME: discuss with FW team: cuda calls are expensive and not needed for each event
```

Comment:
As a side note, this likely need to be per-device, or at least per-device-type (e.g. V100 vs T4 have different limits)